### ISSUE TITLE
Add `applyScaleSet` function

### ISSUE THREAD

**Issue by gegnew** [primitybio/cellengine-r-toolkit#56]
in addition to `applyScale`, we should have a function to apply an entire ScaleSet, like:

```
applyScaleSet = function(scaleset) {
  cols = lapply(1:nrow(scaleset), function (i) {
    channelName = scaleset[i, "channelName"]
    if (channelName %in% colnames(events)) {
      applyScale(scaleset[i, "scale"], events[, channelName])
    }
  })
  df = as.data.frame(do.call(cbind, cols))
  colnames(df) = colnames(events)
  df
}
```

also relevant:

```
col_names = vector(mode="character")
    cols <- vector(mode = "list", length = nrow(scaleset))
    for(i in 1:nrow(scaleset)) {
      channelName = scaleset[i, "channelName"]
      if (channelName %in% colnames(events)) {
        col_names <- append(col_names, channelName)
        cols[[i]] <- applyScale(scaleset[i, "scale"], events[, channelName])
      }
    }
    df = as.data.frame(do.call(cbind, cols))
    colnames(df) = col_names
    out_data = data.matrix(df)
```

**Comment by KevinCodd** [comment]
```
applyScaleSet = function(scaleset, events) {
  cols = lapply(1:length(colnames(events)), function (i) {
    idx <- match(colnames(events)[i], scaleset[,"channelName"], nomatch = NA_integer_)
    if(!is.na(idx)){
      return(applyScale(scaleset[idx, "scale"], events[, colnames(events)[i]]))
    }
    else {
      sprintf("No scale in scaleset for '%s' channel",colnames(events)[i])
      return(events[, colnames(events)[i]])
    }
  })
  df = as.data.frame(do.call(cbind, cols))
  colnames(df) = colnames(events)
  return(df)
}
```
^Probably a decently clean, fast(measured 50% faster than 2nd code snippet on same dataset) and safe way of doing this.

**Comment by gegnew** [comment]
https://rdrr.io/bioc/flowCore/man/transformList-class.html

**Comment by KevinCodd** [comment]
Played with `transformList` and doesn't look as promising as at first glance.  Consider:

```R
scale_funs <- list()
for (c in (colnames(data))) {
  idx <- match(c, scaleset[,"channelName"], nomatch = NA_integer_)
  fn = switch(
    scaleset[idx,"scale"]$type,
    "LinearScale" = linearTransform(transformationId="defaultLinearTransform", a = 1, b = 0),
    "LogScale" = logTransform(transformationId="defaultLogTransform", logbase=10, r=1, d=1),
    "ArcSinhScale" = arcsinhTransform(transformationId="defaultArcsinhTransform", a=1, b=(1/scaleset[idx,"scale"]$cofactor), c=0)
  )
  scale_funs <- append(scale_funs, fn)
}

t1 <- transformList(colnames(data), scale_funs)
outdata <- transform(data, t1)```

Because the scale order in an experiment's scaleset doesn't always match the order in the event columns we still have to create a list of scales to apply based on the column order.  Also it appears that if you pass in a variable as a parameter to the flowCore transform functions, that variable isn't actually evaluated until the transform is applied, thus the above code doesn't actually work because all the cofactors are gotten from the last value of `idx`.  Might not be possible to use this without manually enumerating transforms for each of the different cofactors...maybe I am missing something silly.

**Comment by gegnew** [comment]
I have to look more closely at this, but the reference-catching problem is addressed here: https://win-vector.com/2017/02/26/iteration-and-closures-in-r/

I dealt with it (for a _slightly different purpose: converting a `scaleSet` to a `transformList`) in the `scaleSetToTransformList` in `convertFlowCore.R`

**Comment by gegnew** [comment]
Could you clarify for me exactly what `data` is in this function? I assume `data` is from `flowCore::read.FCS('filename')@exprs` and that `scaleset` is from `getScaleSets(id, fcsFileId)[1,]$scales`

**Comment by gegnew** [comment]
@KevinCodd I've been playing with it, and I _though_ something like the following would work, but it doesn't seem to apply _any_ transform, actually:
```r
scale_funs <- list()
for (c in (colnames(data))) {
  idx <- match(c, scaleset[,"channelName"], nomatch = NA_integer_)
  fn = switch(
    scaleset[idx,"scale"]$type,
    "LinearScale" = linearTransform(transformationId="defaultLinearTransform", a = 2, b = 0),
    "LogScale" = logTransform(transformationId="defaultLogTransform", logbase=10, r=1, d=1),
    "ArcSinhScale" = (function(idx) {
      force(idx)
      return(arcsinhTransform(transformationId="defaultArcsinhTransform", a=0, b=(1/scaleset[idx,"scale"]$cofactor), c=0))
    })(idx)
  )
  scale_funs <- append(scale_funs, fn)
}

t1 <- transformList(colnames(data), scale_funs)
outdata <- transform(data, t1)
```

**Comment by gegnew** [comment]
Alternatively, your `applyScaleSet` function seems to work pretty well

**Comment by KevinCodd** [comment]
@gegnew In the function trying to use the flowCore transforms `data` is just a flowFrame -  `flowCore::read.FCS('filename')`, no `exprs`.  And yes, `scaleset` is `getScaleSets(id, fcsFileId)[1,]$scales`.

**Comment by KevinCodd** [comment]
``scale_funs <- list()
# cofactors <- data.frame
# colnames(cofactors) <- colnames(data)
for (c in (colnames(data))) {
  idx <- match(c, scaleset[,"channelName"], nomatch = NA_integer_)
  fn = switch(
    scaleset[idx,"scale"]$type,
    "LinearScale" = linearTransform(transformationId="defaultLinearTransform", a = 1, b = 0),
    "LogScale" = logTransform(transformationId="defaultLogTransform", logbase=10, r=1, d=1),
    "ArcSinhScale" = arcsinhTransform(transformationId="defaultArcsinhTransform", a=0, b=(1/scaleset[!!idx,"scale"]$cofactor), c=0)
  )
  scale_funs <- append(scale_funs, fn)
}
# scaleset[idx,"scale"]$cofactor))
scale_funs[[1]]
t1 <- transformList(colnames(data), scale_funs)
outdata <- transform(data, t1)``

Okay I was able to make it evaluate with the !! operator.  A majority of the values are just zeroed out though, not sure why yet.

**Comment by zbjornson** [comment]
>  A majority of the values are just zeroed out though, not sure why yet

If you're using a CyTOF dataset, that's normal.

**Comment by KevinCodd** [comment]
Okay that doesn't appear to happen with non-cyTOF datasets, so thank you.  For others though it `NA`s rows (at entries with arcsinh applied) in a recurring pattern.  For example in the Test_Flowcut dataset it follows the pattern: 5 NA, 8 normal, 1 NA, 3 normal, 1 NA, 5 normal.  Patterns are different for different datasets, and some don't appear to have NAs at all.  If you fill in the arcsinh cofactor with a constant e.g. b=1/3, it doesn't NA anything out.  Will look further into this behavior.  

Also, this tentatively looks to be quite a bit faster than the other applyScales method (have anecdotally measured 5-10x) .

**Comment by zbjornson** [comment]
Are you sure you're applying the right scaleset/transformList to the right dataset? I could see that happening if the channel names don't match up.

**Comment by KevinCodd** [comment]
So the issue was actually that `1/scaleset[!!idx,"scale"]$cofactor` was giving back the full cofactor vector, not the cofactor at 'idx'.  The end result was that the transform function repeatedly applied a vector of scales created from the cofactor vector to each channel, explaining the pattern.  

I played around with !!, force(), and eval() quite a bit but still couldn't get it to evaluate.  Finally got it to work with lapply, which I probably should have just used in the first place:

```R
FTapplyScaleSet = function(scaleset, flow_fr) {
  scale_funs <- lapply(colnames(exprs(flow_fr)), function(c) {
    idx <- match(c, scaleset[,"channelName"], nomatch = NA_integer_)
    fn = switch(
      scaleset[idx,"scale"]$type,
      "LinearScale" = linearTransform(transformationId="defaultLinearTransform", a = 1, b = 0),
      "LogScale" = logTransform(transformationId="defaultLogTransform", logbase=10, r=1, d=1),
      "ArcSinhScale" = arcsinhTransform(transformationId="defaultArcsinhTransform", a=0, b=1/(scaleset[idx,"scale"]$cofactor), c=0)
    )
  })
  tl <- transformList(colnames(exprs(flow_fr)), scale_funs)
  outdata <- transform(flow_fr, tl)
  return(exprs(outdata))
}
```   

I think the difference is that lapply forces evaluation of the function arguments when they are added to the output list at the end of each iteration.

This tentatively seems to work, although for _some_ files it gives warnings that NAs are created (have to further investigate).  

Also I timed a run of this vs. the applyScaleSet method in the PR on all of my FCS files, and the FlowCore transform implementation was 13x faster.

BTW this is how I time things - not sure if it is limited or naive:

```R

dirs <- dir(path = "data", pattern = "(FCS)")

num_entries <<- 0
elapsed_time <<- 0

for (d in 1:length(dirs)) {
  #get experiment from directory_name[:-6] -> get scaleset for that experiment
  exp_id <-getExperiment(byName(substring(dirs[d],1,nchar(dirs[d])-6)))["_id"]
  ss <- getScaleSets(exp_id)
  scaleset = ss$scales[[1]]

  files <- grep(list.files(path=paste0("data/",dirs[d]),include.dirs=FALSE), pattern='Compensation Controls|plots|results', invert=TRUE, value=TRUE)
   lapply(files, function(x) {
    print(x)

    flow_fr <- read.FCS(paste0("data/",dirs[d],"/",x), transformation=FALSE)
    flow_mat <- exprs(flow_fr)
    # events <- as.data.frame(flow_mat)

    num_entries <<-num_entries + (nrow(flow_mat)*ncol(flow_mat))
    start.time <- as.numeric(Sys.time())

    df <-applyScaleSet(scaleset, flow_mat)
    # out_data = data.matrix(df)

    end.time <- as.numeric(Sys.time())
    elapsed_time <<- elapsed_time + (end.time - start.time)

  })
}
print(num_entries)
print(elapsed_time)
print(num_entries/elapsed_time)
```

**Comment by zbjornson** [comment]
Nice digging. The version in #58 is invoking the `applyScale` function on each scalar value instead of column vectors because `applyScale()` uses `lapply()`. With that fixed, the method in #58 is slightly faster than flowCore:
```R
> system.time({ applyScaleSet(ss, exprs(fcs)) })
   user  system elapsed 
   0.14    0.00    0.14 
> system.time({ FTapplyScaleSet(ss$scales[[1]], fcs) })
   user  system elapsed 
   0.22    0.00    0.22 
```

**Comment by KevinCodd** [comment]
Ahh makes sense.  All of flowCore's transforms are vectorized, but at the end of the day they just employs the exact same asinh and log functions as in `applyScale`
